P1684-mdarray/P1684.md

```diff
@@ -644,6 +651,440 @@ The larger or equal is intentional, because the mapping may already be not exhau
+with its nodes with keys not in the destination
+(not necessarily all of them)
+transferred from the source (via `extract`)
+to the destination (via `insert` of the node handle).
+
+#### `list`'s `merge` and `splice`
+
+[`list::merge`](https://en.cppreference.com/w/cpp/container/list/merge)
+behaves much like associative containers' `merge`.
+Elements aren't copied or moved;
+instead, their storage is "transferred."
+
+Similarly,
+[`list::splice`](https://en.cppreference.com/w/cpp/container/list/splice)
+transfers elements from the source of the splice to the destination.
+
+#### `basic_string::resize_and_overwrite`
+
+[`basic_string::resize_and_overwrite`](https://en.cppreference.com/w/cpp/string/basic_string/resize_and_overwrite)
+gives callers a way to request and fill uninitialized storage,
+and then hand it off to the string.
+This is like the complement of extraction,
+and is analogous to mdarray's constructors
+that take `container_type&&`.
+
+### Extraction interface options
+
+If WG21 were to require us
+to add this extraction interface to mdarray,
+this section offers suggestions for its design.
+
+#### Suggested names
+
+The name of this operation should convey the following ideas.
+
+* Before the operation, the container belongs to the mdarray.
+
+* The caller reaches inside the mdarray and pulls out the container.
+
+* After the operation, the caller owns the container.
+
+* This "transfer of ownership" works by moving the container.
+
+* The postcondition is an unusable mdarray.
+
+The operation is analogous to smashing a piggy bank
+and pulling coins out of the wreckage.
+
+Here are some possible names, in decreasing order of preference.
+
+1. `extract_container`
+2. `relinquish_container`
+3. `release_container`
+
+Both "relinquish" and "release" convey this idea
+of handing over ownership.
+However, if the container is an `array`
+or a container with similar move behavior,
+then nothing is actually "handed over"
+(except the elements themselves).
+Furthermore, precedence for "release" in `unique_ptr`
+might incorrectly suggest that the extracted-from mdarray
+has a specified state; it does not.
+
+"Extract" suggests that we reach inside the mdarray
+and pull out (move) something that normally belongs there
+(the mdarray owns it beforehand).  After the extraction,
+the caller owns the container.
+The word "extraction" suggests forceful removal
+of a useful component (e.g., a tooth or mineral resources),
+leaving the result depleted or with degraded functionality.
+
+#### Suggested interface
+
+If the interface is a member function, it must have return type `container_type&&` and be nonconst.  Making it rvalue reference qualified might incorrectly suggest that the layout mapping were moved-from.  In fact, the mdarray still has a perfectly valid layout mapping after the operation.
+
+#### Suggested wording
+
+```c++
+container_type&& extract_container();
+```
+
+*Effects*: Equivalent to: `return move(ctr_);`
```

Comment:
We still might need both const and non const version if we do this:
```c++
const container_type&& extract_container() const;
```

Same effects.

Reply:
What would it mean to extract the container from a const mdarray?

Reply:
Would that imply that one could do this?

```c++
template<class E, class L, class C>
auto foo(const mdarray<E, L, C>& m) {
  return mdarray{m.extract_container(), 4, 6};
}

void bar() {
  const mdarray<float, layout_right> m0{2, 3, 4};
  mdarray m1 = foo(m0);
}
```

Reply:
@crtrott I don't think returning `const container_type&&` would make sense here.  The point of `extract_container()` is so that you can move the container out of the mdarray.  You can't move from a const rvalue.